// @antv/graphlib/es/util.js
import { GraphEnum } from './enum';
export function incrementOrInitEntry(map, key) {
  var val = map.get(key) || 0;
  map.set(key, val + 1);
}
export function decrementOrRemoveEntry(map, key) {
  var val = map.get(key);

  if (val !== undefined) {
    val = val - 1;

    if (val > 0) {
      map.set(key, val);
    } else {
      map.delete(key);
    }
  }
}
export function edgeArgsToId(isDirected, v_, w_, name) {
  var v = String(v_);
  var w = String(w_);

  if (!isDirected && v > w) {
    var tmp = v;
    v = w;
    w = tmp;
  }

  return v + GraphEnum.EDGE_KEY_DELIM + w + GraphEnum.EDGE_KEY_DELIM + (name === undefined ? GraphEnum.DEFAULT_EDGE_NAME : name);
}
export function edgeArgsToObj(isDirected, v, w, name) {
  var strV = String(v);
  var strW = String(w);
  var edgeObj = {
    v: v,
    w: w
  };

  if (!isDirected && strV > strW) {
    var tmp = edgeObj.v;
    edgeObj.v = edgeObj.w;
    edgeObj.w = tmp;
  }

  if (name !== undefined) {
    edgeObj.name = name;
  }

  return edgeObj;
}
export function edgeObjToId(isDirected, edgeObj) {
  return edgeArgsToId(isDirected, edgeObj.v, edgeObj.w, edgeObj.name);
}
export function isFunction(target) {
  return !!(target && target.constructor && target.call && target.apply);
}
